### PR TITLE
fix sync reconcilation bug

### PR DESCRIPTION
**Pull request by Frando**
## Description

This fixes two bugs:
* Range contains calculation for iroh-sync compound keys was incorrect
* fs store `get_range` implementation (redb) was incorrect for wrap-around ranges

It also contains these refactorings:
* Improve many_authors test to verify all messages are correct
* Simplify the fs store implementation by adding constructor methods to the range iterators

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

File: iroh-sync/src/net/codec.rs [hunk 1]
```diff
@@ -406,26 +415,35 @@ mod tests {
...
                 );
-
                 assert_eq!(
-                    bob_replica_store
-                        .get(bob_replica.namespace(), GetFilter::all())
-                        .unwrap()
-                        .collect::<Result<Vec<_>>>()
-                        .unwrap()
-                        .len(),
+                    bob_messages.len(),
```
Comment (dignifiedquire) [inline iroh-sync/src/net/codec.rs:446]: We should expand the test to make sure that not just the numbers, but also the actual messages are correct.

File: iroh-sync/src/ranger.rs [hunk 2]
```diff
@@ -568,6 +570,30 @@ where
                     start_index += 1;
                 }
 
+                // TODO: I had troubles understanding the split algorithm above and rewrote it to
```
Comment (dignifiedquire) [inline iroh-sync/src/ranger.rs:573]: please look at the new version from @rklaehn, which is very likely more correct than the implemented one

File: iroh-sync/src/store/fs.rs [hunk 3]
```diff
@@ -444,6 +445,13 @@ impl StoreInstance {
     fn new(namespace: NamespaceId, store: Store) -> Self {
         StoreInstance { namespace, store }
     }
+
+    fn range_start(&self) -> RecordsId {
+        (self.namespace.as_bytes(), &[0u8; 32], &[][..])
+    }
+    fn range_end(&self) -> RecordsId {
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs.rs:452]: these feel super hacky, if anyone has a better idea how to do this, I would love to see it. cc @rklaehn

File: iroh-sync/src/store/fs.rs [hunk 4]
```diff
@@ -634,6 +658,39 @@ pub struct RangeLatestIterator<'a> {
...
+        filter: RangeFilter,
+    ) -> anyhow::Result<Self> {
+        let iter = RangeLatestIterator::try_new(
+            db.begin_read()?,
+            |read_tx| {
+                read_tx
+                    .open_multimap_table(RECORDS_TABLE)
+                    .map_err(anyhow::Error::from)
+            },
+            // |record_table| record_table.range(range).map_err(anyhow::Error::from),
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs.rs:680]: delete?

File: iroh-sync/src/sync.rs [hunk 5]
```diff
@@ -438,13 +463,11 @@ impl PartialOrd for RecordIdentifier {
...
 impl Ord for RecordIdentifier {
     fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        match self.namespace.cmp(&other.namespace) {
-            std::cmp::Ordering::Equal => match self.author.cmp(&other.author) {
-                std::cmp::Ordering::Equal => self.key.cmp(&other.key),
-                res => res,
-            },
-            res => res,
-        }
+        (&self.namespace, &self.author, &self.key).cmp(&(
```
Comment (rklaehn) [inline iroh-sync/src/sync.rs:466]: 👍

File: iroh-sync/src/store/memory.rs [hunk 6]
```diff
@@ -505,11 +505,15 @@ impl crate::ranger::Store<RecordIdentifier, SignedEntry> for ReplicaStoreInstanc
...
         let elements = self.get_range(range.clone(), limit.cloned())?;
         let mut fp = Fingerprint::empty();
+        let mut count = 0;
         for el in elements {
             let el = el?;
             fp ^= el.0.as_fingerprint();
+            count += 1;
+        }
+        if range.x() == range.y() {
+            assert_eq!(count, self.len()?);
```
Comment (dignifiedquire) [inline iroh-sync/src/store/memory.rs:515]: we can remove this

File: iroh-sync/src/sync.rs [hunk 7]
```diff
@@ -390,6 +428,56 @@ pub struct RecordIdentifier {
...
+            "RecId({}:{}:{})",
+            fmt_short(self.namespace.as_bytes()),
+            fmt_short(self.author.as_bytes()),
+            &std::string::String::from_utf8_lossy(&self.key)
+        )
+        // f.debug_struct("RecordIdentifier")
+        //     .field("namespace", &self.namespace)
+        //     .field("author", &self.author)
+        //     .field("key", &std::string::String::from_utf8_lossy(&self.key))
+        //     .finish()
```
Comment (dignifiedquire) [inline iroh-sync/src/sync.rs:451]: delete? change?

**Review by dignifiedquire — APPROVED** [review]
one last nit

File: iroh-sync/src/sync.rs (hunk 7 above)
Comment (Frando) [reply iroh-sync/src/sync.rs]: reverted this for now

**Pull request merged**